packages/core/src/browser/quick-input/quick-view-service.ts

```diff
@@ -30,18 +34,99 @@ export interface QuickViewItem {
 export class QuickViewService implements QuickAccessContribution, QuickAccessProvider {
     static PREFIX = 'view ';
 
-    protected readonly items: (QuickPickItem & { when?: string })[] = [];
+    protected readonly items: (QuickPickItem & {
+        widgetId: string,
+        location: string,
+        defaultLocation: string,
+        viewContainerTitle?: string,
+        when?: string
+    })[] = [];
+    protected readonly containers = new Map<string, { location: string, title: string }>();
     private hiddenItemLabels = new Set<string | undefined>();
+    protected locationsToRestore = new Map<string, string>();
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
 
     @inject(QuickAccessRegistry)
     protected readonly quickAccessRegistry: QuickAccessRegistry;
 
     @inject(ContextKeyService)
-    protected readonly contextKexService: ContextKeyService;
+    protected readonly contextKeyService: ContextKeyService;
+
+    @postConstruct()
+    init(): void {
```

Comment:
A lot of state management has been added here, but I'm not sure it's necessary. When this quick pick is opened, we're dealing with two sets of things: widgets that are already in the shell and widgets that are not already in the shell, but have been registered as openable.

For widgets that are already in the shell, we can just compute everything we need to know about them using utilities on the `ApplicationShell` class.

For widgets that aren't already in the shell, we know where they _would_ go if they were opened, and we can just use that information. We don't need any live state tracking there, either.

My mental model may be a bit naïve, though, so please correct me if I'm missing something.

Reply:
Yes, that would certainly be a nicer solution. However we can't currently get the information we need from `ApplicationShell`. (See also https://github.com/eclipse-theia/theia/pull/10641#issuecomment-1025644454). It would be possible to refactor `ApplicationShell` so it contained more of a model, so `getAreaFor` would get from this model instead of looking through the widget layout structures. I was trying to keep the changes to `ApplicationShell` to a minimum but I am happy to make these changes if you feel that would be better.